<p style="font-family: 'Amiri', serif; font-size: 3.5rem; color: #f8f8f2; text-align: center; margin: 0; text-shadow: 4px 4px 8px rgba(0, 0, 0, 0.6); background: linear-gradient(135deg, #1e1f29, #3a3b4d); padding: 30px; border-radius: 20px; border: 6px solid #bd93f9; width: 95%; letter-spacing: 2px; text-transform: uppercase;">
  🔧 <span style="color: #f1fa8c;">Task 2:</span> <span style="color: #ff79c6;">Data Preprocessing</span> <span style="color: #8be9fd;">& Transformation</span> ⚙️
</p>



<p style="font-family: 'Amiri', serif; font-size: 3rem; color:rgb(23, 119, 122); text-align: center; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #1a1a1a; padding: 20px; border-radius: 20px; border: 7px solidrgb(11, 3, 2); width: 95%;">
  <span style="font-weight: bold; color:rgb(82, 102, 43); animation: pulse 2s infinite;"> Importing   </span>
  <span style="font-style: italic; color:rgb(36, 79, 36); animation: pulse 2s infinite;"> Libraries 📚</span> 
</p>

<style>
@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>


In [40]:
import pandas as pd 
import numpy as np
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
%matplotlib inline 
import warnings 
warnings.filterwarnings('ignore')
import seaborn as sns 
import plotly.express as px
from plotly.offline import iplot
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sqlalchemy import create_engine,text
import re
from mysql import connector
from sqlalchemy import create_engine,text

In [41]:
class Extract_DataBase:
    
    def __init__(self,name_database):
        
        self.name_database=name_database
        
    def extract(self,name_table):
        try:
            engine = create_engine(f'mysql+pymysql://root:2003@localhost:3306/{self.name_database}')
            with engine.connect() as con:
                extract_table = f"SELECT * FROM {name_table}"
                rs = con.execute(text(extract_table))
                data = pd.DataFrame(rs.fetchall(), columns=rs.keys())
            print("Extraction successful.")
            return data
        except Exception as e:
            print(f"Error during extraction: {e}")
            return pd.DataFrame()  

    
if __name__=='__main__':
    extract_data=Extract_DataBase('Bronze_layer')
    data=extract_data.extract('imdb_top_movies')
    if  not data.empty :
        print(data.sample(5))
    else:
        print("None Data")
    
    print('*'*100)
    
    data_crew_cast=extract_data.extract('all_cast_and_crew_with_roles')
    if not data_crew_cast.empty:
         print(data_crew_cast.sample(5))
         
    else:
        print('None Data')
        
    

Extraction successful.
     Movie_ID  Rank                     Title Release Year  Rating  \
211   3315342   212                     Logan       (2017)     8.1   
165   1305806   166  The Secret in Their Eyes       (2009)     8.2   
116    986264   117       Like Stars on Earth       (2007)     8.3   
61      43014    62          Sunset Boulevard       (1950)     8.4   
249     32976   250                   Rebecca       (1940)     8.1   

                Directors                                             Genres  \
211         James Mangold  Dystopian Sci-Fi,Road Trip,Superhero,Action,Dr...   
165  Juan José Campanella  Dark Romance,Erotic Thriller,Legal Thriller,Pe...   
116            Aamir Khan                                       Drama,Family   
61           Billy Wilder  Dark Comedy,Film Noir,Psychological Drama,Show...   
249      Alfred Hitchcock  Dark Romance,Psychological Drama,Psychological...   

     Release Date Country of Origin  \
211          2017     United States 

<p style="font-family: 'Amiri', serif; font-size: 3rem; color: #f0f0f0; text-align: center; margin: 0; text-shadow: 2px 2px 6px rgba(0, 0, 0, 0.6); background-color: #1f2937; padding: 20px; border-radius: 20px; border: 6px solid #38bdf8; width: 95%;">
  <span style="font-weight: bold; color: #facc15; animation: pulse 2s infinite;">Exploratory Data Analysis </span>
  <span style="font-style: italic; color: #34d399; animation: pulse 2s infinite;">for Top 250 IMDB Movies 📊</span> 
</p>

<style>
@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>


In [ ]:
class EDA:
    def __init__(self,data):
        if not isinstance(data,pd.DataFrame):
            raise ValueError("The data parameter should be a valid pandas DataFrame.")
        self.data=data
    
    def display_head(self):
        return (self.data.head(3))
        
    def display_tail(self):
        return(self.data.tail(3))
    
    def display_sample(self):
        return(self.data.sample(3))
        
    def describe_data_numeric(self):
        return self.data.describe()
    
    def describe_data_cat(self):
        return self.data.describe(include=['O'])
    
    def info_data(self):
        """
        Perform Exploratory Data Analysis (EDA) by calculating:
        - Missing values count
        - Percentage of missing values
        - Data types of each feature
        - Number of duplicated values in each feature
        - Number of unique values in each feature
        """
        missing_value_count = self.data.isna().sum()  
        missing_values_Perc = (missing_value_count / len(self.data)) * 100  
        data_types = self.data.dtypes  #  Data types of each feature
        num_dup = self.data.apply(lambda x: x.duplicated().sum())  #  Number of duplicated values per column
        num_uniq = self.data.apply(lambda x: x.nunique())  #  Number of unique values per column
        
        # Concatenate all information into a DataFrame
        info_data = pd.concat(
            [missing_value_count, missing_values_Perc, data_types, num_dup, num_uniq], axis=1
        ).reset_index()
        
        # Rename columns for clarity
        info_data.columns = ['Feature', 'NaN Count', 'NaN Percentage', 'Data Type', 'Num Duplicated', 'Num Unique']
        
        return info_data
        
eda_data=EDA(data)       

In [43]:
#display_head_data
eda_data.display_head()

Movie_ID  Rank                     Title Release Year  Rating  \
0    111161     1  The Shawshank Redemption       (1994)     9.3   
1     68646     2             The Godfather       (1972)     9.2   
2    468569     3           The Dark Knight       (2008)     9.0   

              Directors                                             Genres  \
0        Frank Darabont               Epic,Period Drama,Prison Drama,Drama   
1  Francis Ford Coppola                  Epic,Gangster,Tragedy,Crime,Drama   
2     Christopher Nolan  Action Epic,Epic,Superhero,Tragedy,Action,Crim...   

   Release Date Country of Origin  \
0          1994     United States   
1          1972     United States   
2          2008     United States   

                                      Official Sites              Languages  \
0  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...                English   
1  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...  English,Italian,Latin   
2  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...       English,Mandarin   

                                           Locations  \
0  Mansfield Reformatory - 100 Reformatory Road, ...   
1               Forza d'Agrò, Messina, Sicily, Italy   
2                             Chicago, Illinois, USA   

                                Production Companies Gross Worldwide  
0                          Castle Rock Entertainment     $29,332,133  
1  Paramount Pictures,Albert S. Ruddy Productions...    $250,342,198  
2       Warner Bros.,Legendary Entertainment,Syncopy  $1,009,057,329

In [44]:
#display_tail_data
eda_data.display_tail()

Movie_ID  Rank          Title Release Year  Rating  \
247   1454029   248       The Help       (2011)     8.1   
248    245712   249  Amores Perros       (2000)     8.0   
249     32976   250        Rebecca       (1940)     8.1   

                 Directors                                             Genres  \
247            Tate Taylor                                 Period Drama,Drama   
248  Alejandro G. Iñárritu                             Tragedy,Drama,Thriller   
249       Alfred Hitchcock  Dark Romance,Psychological Drama,Psychological...   

     Release Date Country of Origin  \
247          2011     United States   
248          2000            Mexico   
249          1940     United States   

                                        Official Sites       Languages  \
247  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...         English   
248  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...         Spanish   
249  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...  English,French   

                                             Locations  \
247                        Greenwood, Mississippi, USA   
248  Colonia Condesa, Mexico City, Distrito Federal...   
249                           Big Sur, California, USA   

                                  Production Companies Gross Worldwide  
247  DreamWorks,Dreamworks Pictures,Reliance Film &...    $221,802,186  
248                          Altavista Films,Zeta Film     $20,908,467  
249                    Selznick International Pictures        $113,328

In [45]:
#display_sample_data
eda_data.display_sample()

Movie_ID  Rank                         Title Release Year  Rating  \
94    2106476    95                      The Hunt       (2012)     8.3   
201     15324   202                  Sherlock Jr.       (1924)     8.2   
170     50212   171  The Bridge on the River Kwai       (1957)     8.1   

             Directors                                             Genres  \
94   Thomas Vinterberg                  Psychological Drama,Tragedy,Drama   
201      Buster Keaton  Bumbling Detective,Farce,Feel-Good Romance,Sla...   
170         David Lean      Jungle Adventure,War Epic,Adventure,Drama,War   

     Release Date Country of Origin  \
94           2012           Denmark   
201          1924     United States   
170          1957    United Kingdom   

                                        Official Sites              Languages  \
94   ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...  Danish,English,Polish   
201  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...           None,English   
170  ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...  English,Japanese,Thai   

                                           Locations  \
94            Høje Taastrup Kirke, Taastrup, Denmark   
201  3630 Pasadena Ave, Los Angeles, California, USA   
170                             Ambepussa, Sri Lanka   

                                  Production Companies Gross Worldwide  
94   Zentropa Entertainments,Film i Väst,Zentropa I...     $15,886,373  
201                          Buster Keaton Productions            $399  
170                              Horizon Pictures (II)     $27,201,366

In [46]:
# describe numeric data
eda_data.describe_data_numeric().style.background_gradient('Reds')

In [47]:
# describe object data
eda_data.describe_data_cat()

Title Release Year          Directors  \
count                        250          250                250   
unique                       250           85                155   
top     The Shawshank Redemption       (1995)  Christopher Nolan   
freq                           1            8                  8   

                                   Genres Country of Origin  \
count                                 250               250   
unique                                239                23   
top     Psychological Drama,Tragedy,Drama     United States   
freq                                    4               150   

                                           Official Sites Languages  \
count                                                 250       249   
unique                                                250       108   
top     ['https://help.imdb.com/imdb?ref_=cons_nb_hlp'...   English   
freq                                                    1        73   

                                                Locations  \
count                                                 233   
unique                                                218   
top     Pixar Animation Studios - 1200 Park Avenue, Em...   
freq                                                    7   

                                Production Companies Gross Worldwide  
count                                            250             246  
unique                                           218             246  
top     Walt Disney Pictures,Pixar Animation Studios     $29,332,133  
freq                                               6               1

In [48]:
#Info data 
eda_data.info_data().style.background_gradient('Blues')

<p style="font-family: 'Amiri', serif; font-size: 3rem; color: #e0f2f1; text-align: center; margin: 0; text-shadow: 2px 2px 6px rgba(0, 0, 0, 0.5); background-color: #0f172a; padding: 20px; border-radius: 20px; border: 6px solid #f472b6; width: 95%;">
  <span style="font-weight: bold; color: #60a5fa; animation: pulse 2s infinite;">Exploration</span>
  <span style="font-style: italic; color: #facc15; animation: pulse 2s infinite;">all_cast_and_crew_with_roles 🎭</span> 
</p>

<style>
@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>


In [49]:
eda_data_crew_cest=EDA(data_crew_cast)       

In [50]:
#display_head_data
eda_data_crew_cest.display_head()

Movie_ID           Names       Characters_Position  Rank Primary_Role
0    111161     Tim Robbins             Andy Dufresne   1.0        Actor
1    111161  Morgan Freeman  Ellis Boyd 'Red' Redding   1.0        Actor
2    111161      Bob Gunton             Warden Norton   1.0        Actor

In [51]:
#display_tail_data
eda_data_crew_cest.display_tail()

Movie_ID                 Names      Characters_Position   Rank  \
19031     32976  Dorothy Ann Pailliot              Little Girl  250.0   
19032     32976     Ronald R. Rondell  Hotel Dining Room Guest  250.0   
19033     32976      Phyllis Woodward              Little Girl  250.0   

                                     Primary_Role  
19031                                     Actress  
19032  Second Unit Director or Assistant Director  
19033                                     Actress

In [52]:
#display_sample_data
eda_data_crew_cest.display_sample()

Movie_ID            Names Characters_Position   Rank Primary_Role
91        68646    Lenny Montana          Luca Brasi    2.0        Actor
6744    5311514   Kazuhiko Inoue       Taki's Father   84.0        Actor
18796   5323662  Takuya Masumoto      Keisuke Hirose  245.0        Actor

In [53]:
# describe numeric data
eda_data_crew_cest.describe_data_numeric().style.background_gradient('Reds')

In [54]:
# describe object data
eda_data_crew_cest.describe_data_cat()

Names Characters_Position Primary_Role
count           19034               19034        19010
unique          16498               12346           41
top     Arnold Montey   Additional Voices        Actor
freq               26                 274        12805

In [55]:
#Info data 
eda_data_crew_cest.info_data().style.background_gradient('Oranges')

<p style="font-family: 'Amiri', serif; font-size: 3rem; color: #f0fdfa; text-align: center; margin: 0; text-shadow: 2px 2px 8px rgba(0, 0, 0, 0.4); background-color: #1e1b4b; padding: 20px; border-radius: 20px; border: 6px solid #c084fc; width: 95%;">
  <span style="font-weight: bold; color: #a5f3fc; animation: pulse 2s infinite;">Preprocessing</span>
  <span style="font-style: italic; color: #f472b6; animation: pulse 2s infinite;">imdb_top_movies 🎬</span> 
</p>

<style>
@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>


In [56]:
stand_scaler=StandardScaler()
class Preprocessing_info_top_movies:
    def __init__(self, data):
        # Validate that the input data is a pandas DataFrame
        if not isinstance(data, pd.DataFrame):
            raise ValueError("The data parameter should be a valid pandas DataFrame.")
        self.data = data

    def preprocessing_data(self):
        # Remove parentheses from the 'Release Year' column if they exist
        self.data['Release Year'] = self.data['Release Year'].apply(
            lambda x: x.strip('()') if ')' in x else x
        )

        # Convert 'Release Year' to integer type
        self.data['Release Year'] = self.data['Release Year'].astype(int)

        # Convert 'Release Date' to datetime format and extract the year
        self.data['Release Date'] = pd.to_datetime(self.data['Release Date']).dt.year

        # Clean the 'Gross Worldwide' column by removing non-digit characters and converting to int
        self.data['Gross Worldwide'] = self.data['Gross Worldwide'].apply(
            lambda x: int(re.sub(r'[^\d]', '', x)) 
            if pd.notnull(x) and re.sub(r'[^\d]', '', x) != '' else None
        )

        # Fill missing values in 'Gross Worldwide' with the most frequent value (mode)
        self.data['Gross Worldwide'].fillna(self.data['Gross Worldwide'].mode()[0], inplace=True)

        # Convert the column to Int64 to support nullable integers
        self.data['Gross Worldwide'] = self.data['Gross Worldwide'].astype('Int64')

        # Fill missing values in 'Locations' using values from 'Country of Origin'
        self.data['Locations'] = self.data.apply(
            lambda x: x['Country of Origin'] if pd.isna(x['Locations']) else x['Locations'], axis=1
        )
        #scaler Gross Worldwide
        self.data['Gross Worldwide']=stand_scaler.fit_transform(self.data[['Gross Worldwide']])
        
        return self.data


if __name__ == "__main__":
    # Create a Preprocessing object with the extracted data
    pre_info_data = Preprocessing_info_top_movies(data)

    # Apply the preprocessing steps
    info_data_top_250=pre_info_data.preprocessing_data()
    info_data_top_250.to_csv('imdb_top_movies_transform.csv',index=False)


<p style="font-family: 'Amiri', serif; font-size: 3rem; color: #fefce8; text-align: center; margin: 0; text-shadow: 2px 2px 8px rgba(0, 0, 0, 0.4); background-color: #022c22; padding: 20px; border-radius: 20px; border: 6px solid #4ade80; width: 95%;">
  <span style="font-weight: bold; color: #86efac; animation: pulse 2s infinite;">Preprocessing</span>
  <span style="font-style: italic; color: #fcd34d; animation: pulse 2s infinite;">All Cast & Crew 🎭</span> 
</p>

<style>
@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}
</style>


In [57]:
class Preprocessing_All_Cast_Crew:
    
    def __init__(self,data):
        # Validate that the input data is a pandas DataFrame
        if not isinstance(data, pd.DataFrame):
            raise ValueError("The data parameter should be a valid pandas DataFrame.")
        self.data = data
    def preprocessing_data(self):
        #convert feature RANK FLOAT TO INT
        self.data['Rank']=self.data['Rank'].astype(int)
        
        #handling missing values in feature Primary_Role
        
        self.data['Primary_Role'].fillna('Unknown',inplace=True)
        
        return self.data 
        
if __name__ == "__main__":
    
    # Create a Preprocessing object with the extracted data
    pre_crew_data = Preprocessing_All_Cast_Crew(data_crew_cast)

    # Apply the preprocessing steps
    all_data_crew_cast=pre_crew_data.preprocessing_data()
    all_data_crew_cast.to_csv('all_cast_and_crew_with_roles_transform.csv',index=False)




<p style="font-family: 'Amiri', serif; font-size: 3.5rem; color: #ffffff; text-align: center; margin: 0; text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.6); background: linear-gradient(135deg, #1f4037, #99f2c8); padding: 30px; border-radius: 25px; border: 7px solid #0f9d58; width: 95%; letter-spacing: 1.5px;">
  💾 <span style="color: #ffeb3b;">Database Integration:</span> <span style="color: #ffffff;">Storing Transform Data</span> 📊
</p>
 

In [58]:
class Sliver_layer:
    def __init__(self, data, name_data_base, crew_data=None):
        if not isinstance(data, pd.DataFrame):
            raise ValueError("The data parameter should be a valid pandas DataFrame.")
        self.data = data
        self.name_data_base = name_data_base
        self.crew_data = crew_data

    def integrate_names_and_directors(self):
        if self.crew_data is None:
            raise ValueError("Crew and Cast data (crew_data) must be provided for integration.")
        
        def integrate_names(role_type, df, movie_id):
            names = []
            movie_rows = df[df['Movie_ID'] == movie_id]
            for row in movie_rows.itertuples():
                if row.Primary_Role != role_type:
                    continue
                names.append(row.Names)
            return ', '.join(names) if names else 'Not Available'

        self.data['Actors'] = self.data['Movie_ID'].apply(
            lambda x: integrate_names('Actor', self.crew_data, x)
        )

        self.data['Directors_Extra'] = self.data['Movie_ID'].apply(
            lambda x: integrate_names('Director', self.crew_data, x)
        )

        self.data['Writers'] = self.data['Movie_ID'].apply(
            lambda x: integrate_names('Writer', self.crew_data, x)
        )

        def merge_directors(row):
            parts = []
            if row['Directors_Extra'] == 'Not Available' and row['Writers'] == 'Not Available':
                return row['Directors']
            if row['Directors_Extra'] != 'Not Available':
                parts.append(row['Directors_Extra'])
            if row['Writers'] != 'Not Available':
                parts.append(row['Writers'])
            return ', '.join(parts) if parts else 'Not Available'

        self.data['Directors'] = self.data.apply(merge_directors, axis=1)
        self.data.drop(columns=['Directors_Extra', 'Writers'], inplace=True)
        self.data.to_csv('integrate_Data.csv',index=False)
        print("Integration of actors, directors, and writers completed.")

    def create_data_base(self, user_name, input_host, input_password):
        try:
            with connector.connect(
                host=input_host,
                user=user_name,
                password=input_password
            ) as db:
                cursor = db.cursor()
                cursor.execute(f"CREATE DATABASE IF NOT EXISTS {self.name_data_base}")
                print(f"Database '{self.name_data_base}' created or already exists.")
        except connector.Error as err:
            print(f"Error while creating database: {err}")

    def data_storage(self, name_table):
        try:
            engine = create_engine(f'mysql+pymysql://root:2003@localhost:3306/{self.name_data_base}')
            self.data.to_sql(name=name_table, con=engine, index=False, if_exists='replace')
            print(f"Data stored successfully in table '{name_table}'.")
        except Exception as e:
            print(f"Error while storing data: {e}")
if __name__ == '__main__':
    info_data_top_250 = pd.read_csv(r'D:\assignment data engineering\transform\imdb_top_movies_transform.csv')  
    all_data_crew_cast = pd.read_csv(r'D:\assignment data engineering\transform\all_cast_and_crew_with_roles_transform.csv')  

    name_data_base = 'Sliver_layer'
    name_table = 'integrate_Data_top_movies_transform'

    bronze = Sliver_layer(info_data_top_250, name_data_base, crew_data=all_data_crew_cast)

    bronze.integrate_names_and_directors()

    bronze.create_data_base('root', 'localhost', '2003')
    bronze.data_storage(name_table)



Integration of actors, directors, and writers completed.
Database 'Sliver_layer' created or already exists.
Data stored successfully in table 'integrate_Data_top_movies_transform'.


In [1]:
import pandas as pd

# داتا صناعية فيها طاقم العمل لعدة أفلام
crew_data = pd.DataFrame({
    'Movie_ID': [1, 1, 1, 2, 2, 3],
    'Names': ['Tom Hanks', 'Steven Spielberg', 'John Doe',
              'Will Smith', 'Jane Doe', 'Robert Downey Jr'],
    'Primary_Role': ['Actor', 'Director', 'Writer',
                     'Actor', 'Writer', 'Actor']
})

# نعرض الداتا للتأكد
print(crew_data)


   Movie_ID             Names Primary_Role
0         1         Tom Hanks        Actor
1         1  Steven Spielberg     Director
2         1          John Doe       Writer
3         2        Will Smith        Actor
4         2          Jane Doe       Writer
5         3  Robert Downey Jr        Actor


In [2]:
crew_data 

Movie_ID             Names Primary_Role
0         1         Tom Hanks        Actor
1         1  Steven Spielberg     Director
2         1          John Doe       Writer
3         2        Will Smith        Actor
4         2          Jane Doe       Writer
5         3  Robert Downey Jr        Actor

In [3]:
def integrate_names(role_type, df, movie_id):
    names = []
    movie_rows = df[df['Movie_ID'] == movie_id]  # ناخد فقط الصفوف الخاصة بفيلم محدد
    for row in movie_rows.itertuples():          # نمر على كل صف كـ tuple
        if row.Primary_Role != role_type:        # إذا ما كان الدور المطلوب، تجاهل
            continue
        names.append(row.Names)                 # أضف الاسم
    return ', '.join(names) if names else 'Not Available'


In [10]:
names = []
movie_rows = crew_data[crew_data['Movie_ID'] == 1]  # ناخد فقط الصفوف الخاصة بفيلم محدد
for row in movie_rows.itertuples():          # نمر على كل صف كـ tuple
    if row.Primary_Role != 'Actor': 
            continue
    else :
        print(row.Names)
        
    names.append(row.Names)# إذا ما كان الدور المطلوب، تجاهل 

Tom Hanks


In [23]:
names=[]
movie_id=crew_data[crew_data['Movie_ID']==2]

for row in movie_id.itertuples():
    if row.Primary_Role!='Actor':
        continue
    else:
        names.append(row.Names)
        

In [24]:
names

['Will Smith']